File: memcached/memcached-monitor.py
#!/usr/bin/env python
#-*- coding:utf-8 -*-

__author__ = 'iambocai'

import requests
import json
import time
import socket
import os
import re
import telnetlib
import sys
import commands

class MemcachedStats:

    _client = None
    _key_regex = re.compile(ur'ITEM (.*) \[(.*); (.*)\]')
    _slab_regex = re.compile(ur'STAT items:(.*):number')
    _stat_regex = re.compile(ur"STAT (.*) ([0-9]+\.?[0-9]*)\r")

    def __init__(self, host='localhost', port='11211'):
        self._host = host
        self._port = port

    @property
    def client(self):
        if self._client is None:
            self._client = telnetlib.Telnet(self._host, self._port)
        return self._client

    def command(self, cmd):
        ' Write a command to telnet and return the response '
        self.client.write("%s\n" % cmd)
        return self.client.read_until('END')

    def close(self):
        'close telnet connection'
        return self.client.write("quit\n")

    def key_details(self, sort=True, limit=100):
        ' Return a list of tuples containing keys and details '
        cmd = 'stats cachedump %s %s'
        keys = [key for id in self.slab_ids()
            for key in self._key_regex.findall(self.command(cmd % (id, limit)))]
        if sort:
            return sorted(keys)
        else:
            return keys

    def keys(self, sort=True, limit=100):
        ' Return a list of keys in use '
        return [key[0] for key in self.key_details(sort=sort, limit=limit)]

    def slab_ids(self):
        ' Return a list of slab ids in use '
        return self._slab_regex.findall(self.command('stats items'))

    def stats(self):
        ' Return a dict containing memcached stats '
        return dict(self._stat_regex.findall(self.command('stats')))


def main():
    ip = socket.gethostname()
    timestamp = int(time.time())
    step = 60
    insts_list = [ os.path.basename(i) for i in commands.getoutput(''' ps -ef |grep memcached|grep -v grep |sed -n 's/.* *-p *\([0-9]\{1,5\}\).*/\1/p' ''' ).split('\n') ]
    data = []

    gauges = [ 'get_hit_ratio', 'incr_hit_ratio', 'decr_hit_ratio', 'delete_hit_ratio', 'usage', 'curr_connections', 'total_connections', 'bytes', 'pointer_size', 'uptime', 'limit_maxbytes', 'threads', 'curr_items', 'total_items', 'connection_structures' ]
  
    for inst in insts_list:
        
        port = inst
        metric = "memcached"
        endpoint = ip
        tags = 'port=%s' % port

        try:
            # ATT: if your instance listened on 127.0.0.1, change "ip" to "127.0.0.1" in follow line
            conn = MemcachedStats(ip, port)
            stats = conn.stats()
            conn.close()
        except:
            continue

        del stats['pid']
        del stats['time']

        stats['usage'] = str(100 * float(stats['bytes']) / float(stats['limit_maxbytes']))
        try:
            stats['get_hit_ratio'] = str(100 * float(stats['get_hits']) / (float(stats['get_hits']) + float(stats['get_misses'])))
        except ZeroDivisionError:
            stats['get_hit_ratio'] = '0.0'
        try:
            stats['incr_hit_ratio'] = str(100 * float(stats['incr_hits']) / (float(stats['incr_hits']) + float(stats['incr_misses'])))
        except ZeroDivisionError:
            stats['incr_hit_ratio'] = '0.0'
        try:
            stats['decr_hit_ratio'] = str(100 * float(stats['decr_hits']) / (float(stats['decr_hits']) + float(stats['decr_misses'])))
        except ZeroDivisionError:
            stats['decr_hit_ratio'] = '0.0'
        try:
            stats['delete_hit_ratio'] = str(100 * float(stats['delete_hits']) / (float(stats['delete_hits']) + float(stats['delete_misses'])))
        except ZeroDivisionError:
            stats['delete_hit_ratio'] = '0.0'


        for key in stats:
            value  = float(stats[key])
            if key in gauges:
                suffix = ''
                vtype  = 'GAUGE'
            else:
                suffix = '_cps'
                vtype = 'COUNTER'

            i = {
                'metric': '%s.%s%s' % (metric, key, suffix),
                'endpoint': endpoint,
                'timestamp': timestamp,
                'step': step,
                'value': value,
                'counterType': vtype,
                'tags': tags
            }
            data.append(i)

    return data
    

if __name__ == '__main__':
    proc = commands.getoutput(''' ps -ef|grep 'memcached-monitor.py'|grep -v grep|wc -l ''')
    if int(proc) < 3:
        r = requests.post("http://127.0.0.1:1988/v1/push", data=json.dumps(main()))
        print r.text

File: memcached/memcached_monitor.php
<?php
//php版本memcached监控
$falArr = $data = array();
$conTimeout = 1;
//采集数据
if (extension_loaded('memcached')) {
    $config = array(
        'servers' => array(//服务器组
            array('127.0.0.1', 11211),
        ),
        'policy' => array(//缓存策略
            'compressed' => false, //是否压缩,开启后，默认大于100b时压缩
            'life_time' => 3600, //缓存时间 秒
            'pool' => 'mc', //连接池，同一连接池建立长连接
        )
    );
    $memcached = new \Memcached($config['policy']['pool']);
    $memcached->setOption(\Memcached::OPT_COMPRESSION, $config['policy']['compressed']);
    $memcached->setOption(\Memcached::OPT_CONNECT_TIMEOUT, $conTimeout);
    if ($memcached->addServers($config['servers'])) {
        $data = $memcached->getStats();
    }
}
//要监控的key列表
$monitKeys = array(
    'uptime' => 'GAUGE',
    'threads' => 'GAUGE',
    'pointer_size' => 'GAUGE',
    'rusage_user_seconds' => 'GAUGE',
    'rusage_user_microseconds' => 'GAUGE',
    'rusage_system_seconds' => 'GAUGE',
    'rusage_system_microseconds' => 'GAUGE',
    'curr_items' => 'GAUGE',
    'total_items' => 'GAUGE',
    'limit_maxbytes' => 'GAUGE',
    'curr_connections' => 'GAUGE',
    'total_connections' => 'GAUGE',
    'connection_structures' => 'GAUGE',
    'bytes' => 'GAUGE',
    'cmd_get' => 'GAUGE',
    'cmd_set' => 'GAUGE',
    'get_hits' => 'GAUGE',
    'get_misses' => 'GAUGE',
    'evictions' => 'GAUGE',
    'bytes_read' => 'GAUGE',
    'bytes_written' => 'GAUGE',
);
//关联监控指标数值
foreach ($monitKeys as $k => $v) {
    $falArr[] = array(
        "endpoint" => gethostname(),
        "metric" => "memcached_" . $k,
        "timestamp" => time(),
        "step" => 60,
        "value" => (isset($data[$k]) ? $data[$k] : 0),
        "counterType" => $v,
        "tags" => "project=atido,group=server,name=memcached,method=monitor",
    );
}
//stdOut
echo json_encode($falArr);
